tiled/_tests/test_authentication.py

```diff
@@ -94,6 +94,41 @@ def test_password_auth(enter_username_password, config):
                 from_context(context, username="alice")
 
 
+def test_password_auth_hook(config):
+    with Context.from_app(build_app_from_config(config)) as context:
+        # Passing prompt_for_reauthentication=False should riase an
```

Comment:
```suggestion
        # Passing prompt_for_reauthentication=False should raise
```

Reply:
Unrelated to my copy-editing above, I have questions about the desired behavior.

I had in mind that `context.authenticate(username, password, prompt_for_reauthication=False)` should _not_ raise immediately, but rather configure the `context` to raise _later_ if the session expires and fresh credentials are required. The idea was that the graphical application would catch this error.

Reflecting on this now, I am questioning whether exceptions are the best signaling mechanism. Philosophically, is not an "exceptional" situation: it's a very foreseeable one. And pragmatically, forcing the application to carry every client interaction inside a `try:` is inelegant.

What if we had these options:
- `True` -> current behavior
- `False` -> raise `CannotPrompt` if creds expire
- `callable` -> callback hook that graphical application can use to pop up a dialog or something for new creds

Reply:
This makes more sense. I'll implement something to that effect. I'll want to chat through the workflow a little bit to understand when/how reauth is normally triggered, but I think I can get the right bones in place.

Reply:
I followed through on this in https://github.com/bluesky/tiled/pull/816/commits/a9e97c0eca17d0cf89ece5925173b3cb90a79631.


And updated the tests in 
https://github.com/bluesky/tiled/pull/816/commits/877ff19b335d0b88ca34c9812adc4151af6220e4

Worth noting that this removes the raising of `CannotRefreshAuthentication`, just removes the `http_client.auth`, and allows that to be caught at `CannotPrompt`.